src/box/decimal.h

```diff
@@ -0,0 +1,574 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2022, Tarantool AUTHORS, please see AUTHORS file.
+ */
+#pragma once
+
+#include "trivia/util.h"
+
+#if defined(__cplusplus)
+extern "C" {
+#endif /* defined(__cplusplus) */
+
+/** \cond public */
+
+/* {{{ decimal structure and constants */
+
+/** See @ref box_decimal_t. */
+struct box_decimal {
+	/**
+	 * Decimal data.
+	 *
+	 * The format is unspecified and can be changed.
+	 *
+	 * The tail of the array may be in an unallocated memory.
+	 * Don't access this array directly. Use box_decimal_*()
+	 * functions instead.
+	 */
+	uint64_t internal[8];
+};
+
+/**
+ * Storage for a decimal number.
+ *
+ * This type is large enough to hold a decimal value. The primary
+ * usage is to allocate a decimal on the stack and pass it to a
+ * box_decimal_*() function.
+ *
+ * Take it as opacue structure with ability to allocate a value
+ * on the stack.
+ *
+ * Use box_decimal_copy() to copy the value, don't use memcpy().
+ * The real data can be smaller than this type. Moreover,
+ * tarantool may allocate less bytes for @ref box_decimal_t value,
+ * so direct read/write of the structure may lead to access to an
+ * unallocated memory.
+ *
+ * The alignment of the structure is not less than alignment of
+ * decimal values allocated by tarantool. It can be larger and it
+ * is another reason to don't access the structure content
+ * directly.
+ *
+ * Last, but not least: the data format is unspecified and can be
+ * changed. Modules and applications that aims to be compatible
+ * with future tarantool versions must not lean on particular data
+ * representation within the structure.
+ */
+typedef struct box_decimal box_decimal_t;
+
+/**
+ * Buffer of this size is enough to hold any
+ * box_decimal_to_string() result.
+ */
+#define BOX_DECIMAL_STRING_BUFFER_SIZE 128
+
+/* }}} decimal structure and constants */
+
+/* {{{ decimal dynamic allocation (helpers) */
+
+/**
+ * Amount of bytes enough to hold any decimal value.
+ *
+ * This value is less or equal to sizeof(box_decimal_t). It is
+ * valid only for particular tarantool version. The value is a
+ * runtime constant: it is not changed during tarantool execution.
+ *
+ * The function is suitable for dynamic allocation of a storage
+ * for a decimal type, especially when the memory is managed by
+ * a runtime of another language.
+ *
+ * @return max decimal size in bytes on current tarantool version
+ */
+API_EXPORT size_t
+box_decimal_storage_size(void);
+
+/**
+ * Alignment of the decimal storage.
+ *
+ * It is less or equal to alignof(box_decimal_t). The value is
+ * valid only for particular tarantool version. The value is a
+ * runtime constant: it is not changed during tarantool execution.
+ *
+ * @return decimal type alignment on current tarantool version
+ */
+API_EXPORT size_t
+box_decimal_storage_align(void);
```

Comment:
I don't think we should provide the size and alignment functions. A decimal will be often allocated as a part of an array or a struct, in which case they are useless. I doubt module developers will manage memory directly just to save a few bytes per decimal. Let's add these function only if somebody asks for them.

Reply:
('A few bytes per decimal' is almost half of the decimal size: 36 vs 64.)

It is already allocated dynamically in [tarantool-module](https://github.com/picodata/tarantool-module) (a module that implements bindings for Rust). See [here](https://github.com/picodata/tarantool-module/blob/597bd399b225eacfcacfbfae09fb8df6c952c2db/tarantool/src/decimal.rs#L361) for example.

I discussed it with @gmoshkin and he said that he is not interested in storage_size, storage_align functions and will hardcode their values in the Rust module code (and, it seems, will update them in sync with tarantool versions). AFAIU, there are some optimization plans, which anyway mean that the Rust module will lean on the internal data layout. (That's a bit sad for me.)

I think it would be good to give ability to reduce memory footprint for users, which anyway allocate decimals dynamically: either with an own allocator or by using a runtime of another language. However, well, it is not critical to solve our current task and can be at least postponed.

I'll drop those functions from this pull request.

Reply:
Dropped the following functions:

- `box_decimal_storage_size()`
- `box_decimal_storage_align()`
- `box_decimal_string_buffer_size()`